android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java

```diff
@@ -1967,15 +1995,36 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
         editor.putBoolean(KMManager.KMKey_ShouldShowHelpBubble, false);
         editor.commit();
 
+        // Update globeKeyState
+        int start = url.indexOf("keydown=") + 8;
+        String value = url.substring(start);
+        boolean globeKeyDown = !value.isEmpty() && Boolean.valueOf(value);
+        if (globeKeyState != GlobeKeyState.GLOBE_KEY_STATE_LONGPRESS) {
+          if (globeKeyDown) {
+            globeKeyState = GlobeKeyState.GLOBE_KEY_STATE_DOWN;
+          } else {
+            globeKeyState = GlobeKeyState.GLOBE_KEY_STATE_UP;
+          }
+        }
+
         if (KMManager.shouldAllowSetKeyboard()) {
-          if (InAppKeyboard.keyboardPickerEnabled) {
-            if (inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SHOW_MENU) {
+          if (globeKeyState == GlobeKeyState.GLOBE_KEY_STATE_LONGPRESS) {
+            // Longpress globe
+            if (InAppKeyboard.keyboardPickerEnabled) {
               showKeyboardPicker(context, KeyboardType.KEYBOARD_TYPE_INAPP);
-            } else if (inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SWITCH_TO_NEXT_KEYBOARD) {
+              globeKeyState = GlobeKeyState.GLOBE_KEY_STATE_UP;
+            }
+          } else if (globeKeyState == GlobeKeyState.GLOBE_KEY_STATE_UP) {
+            // Handle shortpress globe
+            if (InAppKeyboard.keyboardPickerEnabled) {
+              if (inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SHOW_MENU) {
+                showKeyboardPicker(context, KeyboardType.KEYBOARD_TYPE_INAPP);
+              } else if (inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SWITCH_TO_NEXT_KEYBOARD) {
+                switchToNextKeyboard(context);
+              }
+            } else {
               switchToNextKeyboard(context);
             }
```

Comment:
This would be easier to follow if we extracted out the actions from the tests, for example:


```java
          if (globeKeyState == GlobeKeyState.GLOBE_KEY_STATE_LONGPRESS) {
            // Longpress globe
            doGlobeKeyLongpressAction();
            globeKeyState = GlobeKeyState.GLOBE_KEY_STATE_UP;
          } else if (globeKeyState == GlobeKeyState.GLOBE_KEY_STATE_UP) {
            // Handle shortpress globe
            doGlobeKeyShortpressAction();
          }
```

These could be refactored slightly more (plus we have the whole in-app vs system keyboard mess -- one of these days this needs cleanup), but as a start:

```java
  void doGlobeKeyLongpressAction() {
    if (InAppKeyboard.keyboardPickerEnabled) {
      showKeyboardPicker(context, KeyboardType.KEYBOARD_TYPE_INAPP);
    }
  }

  void doGlobeKeyShortpressAction() {
    if (InAppKeyboard.keyboardPickerEnabled) {
      if (inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SHOW_MENU) {
        showKeyboardPicker(context, KeyboardType.KEYBOARD_TYPE_INAPP);
      } else if (inappKbGlobeKeyAction == GlobeKeyAction.GLOBE_KEY_ACTION_SWITCH_TO_NEXT_KEYBOARD) {
        switchToNextKeyboard(context);
      }
    } else {
      switchToNextKeyboard(context);
    }
  }
```

Reply:
I did the refactoring. For now, the doGlobeKey*() functions only handle IN_APP keyboard.
the System Keyboard is complicated by the lock-screen code.